### ISSUE TITLE
Review internal functions intersectpoint, intersectcollinear, ...

### ISSUE THREAD

**Issue by juliohm** [JuliaGeometry/Meshes.jl#299]
Some of these internal functions are now superseded by the new `intersectparameters` by @dorn-gerhard . We should review and probably remove them from the code base.

**Comment by juliohm** [comment]
@dorn-gerhard did you have a chance to take a look into this?

**Comment by dorn-gerhard** [comment]
I was thinking of generalizing intersectparameters to define intersections of any two affine subvectorspaces, e.g. two planes in a 3, 4, or n-dimensional space. 
This could be used to prepare intersections of lines with planes (thinking of using this as basis for triangle-ray intersection) and so on.

It boils all down to defining the subvectorspace through a set of points (function input could be two vectors of points) and leads to the same linear systems to solve and the same rank analysis (using qr decomposition and so on).
function could look like:
``` math
λ⃗₁, λ⃗₂, r, rₐ = intersectparameters(p₁::Vector{Point{N,T}}, p₂::Vector{Point{N,T}})
```

With ``N`` the dimension of the space, ``n₁,n₂ = length(p₁)-1, length(p₂)-1`` the dimensions of the affine subvectorspaces, one has the following possibilities for the ranks:

$\max(n_1, n_2) \leq r \leq \max(N,n_1 + n_2)$
 $\max(n_1, n_2) \leq r \leq r_a \leq r+1 \leq \max(N,n_1 + n_2 + 1)$

If ranks are equal there is a solution, the dimension of the resulting intersection subvectorspace $n_i$ is given by
$n_i = N - r$ ($r =r_a$)

Example:
3D, 1 line given by p₁, 1 plane given by p₂, $N= 3, n_1 = 1, n_2 = 2$,
$r \in [2,3], r_a \in [2,3]$
Cases:
$r = 3, r_a = 3$: Point (lambdas are well defined),  (line p₁ hits plane p₂)
$r = 2, r_a = 3$: no intersection  (line p₁ parallel to plane p₂)
$r = 2, r_a = 2$: Line (lambda can be set to zero), (line p₁ lies within plane p₂)

Example:
6D,  affine 3D subspace given by p₁, plane given by p₂, $N= 6, n_1 = 3, n_2 = 2$,
$r \in [3,4,5], r_a \in [3,4,5,6]$
Cases:
$r = 3, r_a = 3$: Plane (lambdas can be set to zero),  (plane p₂ lies within 3D p₁)
$r = 3, r_a = 4$: no intersection  (2 vectors of 3D p₁ parallel to plane p₂)
$r = 4, r_a = 4$: Line (lambda not well defined), (3D p₁ intersects plane p₂ in one line)
$r = 4, r_a = 5$: no intersection  (1 vector of 3D p₁ and of plane p₂ is parallel)
$r = 5, r_a = 5$: Point (lambdas are well defined), (3D p₁ intersects plane p₂ in one point)
$r = 5, r_a = 6$: no intersection (lambdas indicate points of shortest distance), (3D p₁ and plane p₂ are skew)

I have one remaining question for implementing this general function:
- What to return for lambdas (they generally represent a point) if the intersecting subvectorspace is for instance 2 dimensional.
The original idea of the lambdas was to show how to reach the intersecting point (now subspace) from both input subvectorspaces.
One solution could be: returning two vectors of length $n_i+1$ of lambda vectors [λ⃗¹,λ⃗²,λ⃗³]₁, [λ⃗¹,λ⃗²,λ⃗³]₂ to describe the intersecting subvectorspace.
in case of two intersecting planes in 3D this could be:
[λ⃗¹,λ⃗²]₁, [λ⃗¹,λ⃗²]₂ with each lambda vector having two components.

-> but this would break type stability

- Alternative would be to indicate to a nearest point of the subvectorspace.

What do you think?

**Comment by juliohm** [comment]
@dorn-gerhard I am not sure I am following all the details but I trust you are considering all the downsides of different designs. Why returning vectors of lambdas would be type unstable? Wouldn't that always be the return type in all cases?

**Comment by dorn-gerhard** [comment]
Regarding **design**: I wanted to prevent to have several functions if one function could do everything.
I started thinking to discuss line - plane intersection and generalized it to any intersection of affine subvectorspace - i think the design is quite appealing.

Main design ideas are:
- representation of affine subvectorspaces by a set of points (contradicts a bit the primitive plane, which uses already vectors, but more general for and (sub)dimension)
- using the same arguments (namely the ranks $r$ and $r_a$) about the solvability of linear systems to decide the intersection question, implemented by a clever qr decomposition
- return a lambda to indicate for each subvectorspace how to reach the intersection subvectorspace
  1) a number: subvectorspace is a line, intersection is a point, e.g. ``λ₁= 3.4``
  2) a vector: subvectorspace is a (hyper)plane, intersection a point, e.g. ``λ₁ = [2.1, 3.4]``
  3) a vector of vectors: subvectorspace is e.g. a (hyper)plane, intersection is at least a line or higher dimensional, e.g. ``λ₁ = [[2.1, 3.1, -0.1], [1.1, 3.2, 3.2]]``

Regarding **type stability**:
So I know from the input types how long the return vectors lambda will be. What I cannot know beforehand is the dimensionality of the intersection (could be Point (standard case), but also Line or Plane)
In order to describe them I would need a vector of vectors for each lambda.
I guess this contradicts type stability which is the reason why I suggest to return only a single point (one closest to the intersection), if the dimensionality of the intersection is larger than 0.

So for the examples before we would get:
1) ``λ₁= 3.4``
2) ``λ₁ = [2.1, 3.4]``
3) ``λ₁ = [2.1, 3.1, -0.1]`` (one nearest point)

in the case of no intersection, the numbers would be just set to zero.

The change in design would be:
instead of 
``λ₁::T, λ₂::T , r::Int64, rₐ::Int64 = intersectparameters(a::Point{Dim,T}, b::Point{Dim,T}, 
                             c::Point{Dim,T}, d::Point{Dim,T})``
we would have
``λ₁::Any{T,Vector{T}}, λ₂::Any{T,Vector{T}}, r::Int64, rₐ::Int64 = intersectparameters(p₁::Vector{Point{Dim,T}}, p₂::Vector{Point{Dim,T}}``

return type (T or Vector{T}) would be known from the size of the inputs
length(p₁) > 2 -> λ₁::Vector{T}
length(p₁) == 2 -> λ₁::T
What do you think?

**Comment by juliohm** [comment]
@dorn-gerhard can you please confirm that we need to decide on this design now in order to fix the issue? I thought that the fix to the current issue was orthogonal to the design of the new intersectparams? If you can replace the intersectpoint and intersectcollinear, ... with the current implementation of intersectparams I will be able to provide more helpful feedback.

**Comment by dorn-gerhard** [comment]
ah okay, you are right, I should focus on the issue topic.
 
So, currently the two functions just appear in 
`intersectpoint`:
- ``segments.jl`` (defined for segments, used - but obsolete)
- ``raysegment.jl`` (defined, only used in tests)
- ``rays.jl`` (only defined for rays, never used``)
- ``lines.jl`` (defined for lines, used in the definition of the other implementations)
- ``test/intersections.jl`` (only used for one test (raysegment)
 
`intersectcollinear`:
- ``segments.jl`` (defined and used - i think obsolete

From my point of view both functions can be replaced by ``intersectparameters``

Biggest changes have to be done in ``segments.jl``
I will also provide a benchmark to see if the 2D version is more efficient than my suggested method.

I will open a PR with all needed changes the next days.

**Comment by dorn-gerhard** [comment]
Comment: 
intersectcollinear has a bug:
``` julia
s1 = Segment((0,0), (1,-1))
s2 = Segment((0.5,-0.5), (1.5,-1.5))
Meshes.intersectcollinear(identity, s2, s1)
```
yields
``` julia
Meshes.Segment{2,Float64}
  └─Point(1.5, -1.0)
  └─Point(0.0, -0.5)
```
instead of 
``` julia
Meshes.Segment{2,Float64}
  └─Point(1.0, -1.0)
  └─Point(0.5, -0.5)
```

I will repair that.

**Comment by juliohm** [comment]
Fixed by #309
